src/Features/Core/Portable/UnifiedSuggestions/UnifiedSuggestedActionsSource.cs

```diff
@@ -0,0 +1,711 @@
+            => new UnifiedSuggestedActionSet(set.CategoryName, set.Actions, set.Title, priority, set.ApplicableToSpan);
+
+        private static ImmutableArray<UnifiedSuggestedActionSet> InlineActionSetsIfDesirable(
+            ImmutableArray<UnifiedSuggestedActionSet> allActionSets)
+        {
+            // If we only have a single set of items, and that set only has three max suggestion 
+            // offered. Then we can consider inlining any nested actions into the top level list.
+            // (but we only do this if the parent of the nested actions isn't invokable itself).
+            if (allActionSets.Sum(a => a.Actions.Count()) > 3)
+            {
+                return allActionSets;
+            }
+
+            return allActionSets.SelectAsArray(InlineActions);
+        }
+
+        private static UnifiedSuggestedActionSet InlineActions(UnifiedSuggestedActionSet actionSet)
+        {
+            using var newActionsDisposer = ArrayBuilder<IUnifiedSuggestedAction>.GetInstance(out var newActions);
+            foreach (var action in actionSet.Actions)
+            {
+                var actionWithNestedActions = action as UnifiedSuggestedActionWithNestedActions;
+
+                // Only inline if the underlying code action allows it.
+                if (actionWithNestedActions?.CodeAction.IsInlinable == true)
+                {
+                    newActions.AddRange(actionWithNestedActions.NestedActionSets.SelectMany(set => set.Actions));
+                }
+                else
+                {
+                    newActions.Add(action);
+                }
+            }
+
+            return new UnifiedSuggestedActionSet(
+                actionSet.CategoryName,
+                newActions.ToImmutable(),
+                actionSet.Title,
+                actionSet.Priority,
+                actionSet.ApplicableToSpan);
+        }
+
+        private static ImmutableArray<UnifiedSuggestedActionSet> FilterActionSetsByTitle(
+            ImmutableArray<UnifiedSuggestedActionSet> allActionSets)
+        {
+            using var resultDisposer = ArrayBuilder<UnifiedSuggestedActionSet>.GetInstance(out var result);
+            var seenTitles = new HashSet<string>();
+
+            foreach (var set in allActionSets)
+            {
+                var filteredSet = FilterActionSetByTitle(set, seenTitles);
+                if (filteredSet != null)
+                {
+                    result.Add(filteredSet);
+                }
+            }
+
+            return result.ToImmutable();
+        }
+
+        private static UnifiedSuggestedActionSet? FilterActionSetByTitle(UnifiedSuggestedActionSet set, HashSet<string> seenTitles)
+        {
+            using var actionsDisposer = ArrayBuilder<IUnifiedSuggestedAction>.GetInstance(out var actions);
+
+            foreach (var action in set.Actions)
+            {
+                if (seenTitles.Add(action.CodeAction.Title))
```

Comment:
@mavasani I changed the logic in this line [from using DisplayText](http://sourceroslyn.io/#Microsoft.CodeAnalysis.EditorFeatures.Wpf/Suggestions/SuggestedActionsSource.cs,309) to instead using the title of the code action, since adding the DisplayText property to the unified suggested action class proved to be difficult due to some classes being unavailable in the Features layer. I tested and things seemed to work fine, but would this change potentially introduce problems?

Reply:
this is a good question, and i would like @mavasani to weigh in

Reply:
I talked to Manish and he says this change should likely be fine, so will mark this comment as resolved.